CU-8678g5b7w - Create an article on neo website explaining how to use… (#3)
CityOfZion/neo-dev-portal: tutorials/2023-11-06-offchain-sdk-ts-generation-with-cpm/index.md

- `contracts`: will have the Neo Token contract that we want to generate;
- `tools`: since we don't want to download the smart contract, this section is unnecessary;
- `networks`: will have the Main Net network where the Neo Token contract is.

After tweaking the `cpm.yaml` file it should be something like this:

```yaml
defaults:
  contract-source-network: mainnet
  contract-generate-sdk: true
  contract-download: false
  off-chain:
    languages:
    - ts
    destinations:
	    ts: src/contracts

contracts:
  - label: NEO token
    script-hash: '0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5'

networks:
  - label: mainnet
    hosts:
      - 'https://mainnet1.neo.coz.io:443'
      - 'http://seed1.neo.org:10332'
```

> For more information on how to configure `cpm.yaml`, check out [CPM's documentation](https://github.com/CityOfZion/cpm/blob/master/docs/config.md)

Then just run `cpm run` on the terminal and the SDK should be generated at the `src/contracts/neotoken` folder.' Anytime the contract adds or removes method signatures, you'll have to run `cpm run` again to update your SDK.

## 3. Initializing and running the SDK

> Now that the SDK is already generated, you'll need to install the dependencies: [`@types/node`](https://www.npmjs.com/package/@types/node)], and  [`@cityofzion/neon-dappkit-types`](https://www.npmjs.com/package/@cityofzion/neon-dappkit-types). Then you'll have to import and initialize the SDK to be able to call it.

Reviewer: There is a lost `]` in the text
```suggestion
Now that the SDK is already generated, you'll need to install the dependencies: [`@types/node`](https://www.npmjs.com/package/@types/node), and  [`@cityofzion/neon-dappkit-types`](https://www.npmjs.com/package/@cityofzion/neon-dappkit-types). Then you'll have to import and initialize the SDK to be able to call it.
```

> To initialize the SDK, you'll first need to instantiate a class that implements `Neo3-Invoker`, like [`NeonInvoker`](https://github.com/CityOfZion/neon-dappkit/blob/main/packages/neon-dappkit/src/NeonInvoker.ts) or [`WalletConnect SDK`](https://github.com/CityOfZion/wallet-connect-sdk/tree/main). If you plan to allow users to log into their accounts and make them do the invocation on your project, then Wallet Connect is the better option, otherwise, NeonInvoker can work too.  

Reviewer: ```suggestion
To initialize the SDK, you'll first need to instantiate a class that implements `Neo3-Invoker`. 
To enhance security and user experience, it's advisable to use [`WalletConnect`](https://github.com/CityOfZion/wallet-connect-sdk) for scenarios requiring user authentication, such as signing transactions or accessing user account details. However, for situations where such authentication is unnecessary, [`NeonInvoker`](https://github.com/CityOfZion/neon-dappkit/blob/main/packages/neon-dappkit/src/NeonInvoker.ts) from NeonDappkit offers a more streamlined solution, as it eliminates the need for users to approve requests through their wallets. In our example, we employ both methods: NeonInvoker is utilized when the user is not logged into their wallet.
```

```ts
import { NeoToken } from './contracts/neotoken'

import { NeonInvoker } from "@cityofzion/neon-dappkit"
import WcSdk from '@cityofzion/wallet-connect-sdk-core'
import SignClient from '@walletconnect/sign-client'

async function createNeoTokenSDKWithWcSDK(): Promise<NeoToken> {

> 	let wcSdk
> 	let signClient
> 
> 	signClient = await SignClient.init({

Reviewer: The indentation is strange
```suggestion
    let wcSdk
    let signClient

    signClient = await SignClient.init({
```

        projectId: 'a9ff54e3d56a52230ed8767db4d4a810', // the ID of your project on Wallet Connect website
        relayUrl: 'wss://relay.walletconnect.com', // we are using walletconnect's official relay server
        metadata: {
          name: 'MyApplicationName', // your application name to be displayed on the wallet
          description: 'My Application description', // description to be shown on the wallet
          url: 'https://myapplicationdescription.app/', // url to be linked on the wallet
          icons: ['https://myapplicationdescription.app/myappicon.png'], // icon to be shown on the wallet
        },
    })
    wcSdk = new WcSdk(signClient)

    return new NeoToken({
		scriptHash: NeoToken.SCRIPT_HASH,
		invoker: wcSdk
	})

> }

Reviewer: I just noticed there is a problem on this example: a new wc instance will be created everytime the user want's to get the SDK. Let's try something more similar to this: https://github.com/CityOfZion/neon-dappkit/blob/main/packages/neon-dappkit/WALLET-CONNECT.md

async function createNeoTokenSDKWithNeonInvoker(): Promise<NeoToken> {
	return new NeoToken({
		scriptHash: NeoToken.SCRIPT_HASH,
		invoker: await NeonInvoker.init({

> 			rpcAddress: 'https://mainnet1.neo.coz.io:443',

Reviewer: This is easier to remember
```suggestion
			rpcAddress: NeonInvoker.MAINNET,
```

		})
	})
}
```

Then you can invoke the contract methods with this SDK:
```ts
async main() {

> 	const neoToken = await createNeoTokenSDK()

Reviewer: You didn't create this method, and I think we should do something like this: https://github.com/CityOfZion/neon-dappkit/blob/main/packages/neon-dappkit/WALLET-CONNECT.md?plain=1#L19

Reviewer: Put a text somewhere near this method saying we are not teaching how to connect, to do so, the reader can check wcsdk documentation.

---
title: 'How to use CPM to generate an off-chain TypeScript SDK'
description: "The tutorial aims to provide a comprehensive guide for developers interested in creating TypeScript SDKs for Neo smart contracts, making it easier for them to leverage CPM and successfully work with off-chain SDKs."
author: coz.io
tags: ["TS", "CPM", "SDK"]
skill: BEGINNER
sidebar: true
---

[CPM](https://github.com/CityOfZion/cpm) is a tool that can help you download smart contracts and generate on-chain and off-chain SDKs for multiple languages without the need of writing them yourself and needing none or minimal changes.

> ## 1. Prerequisites

Reviewer: I think we should get rid of the "Prerequisites" section. It would be better to guide the user step by step creating a clean new project that uses CPM. (use vite to create the project)

You need to have [CPM installed](https://github.com/CityOfZion/cpm/tree/master#installation), and your TypeScript project will need to have the [`@types/node`](https://www.npmjs.com/package/@types/node) and [`@cityofzion/neon-dappkit-types`](https://www.npmjs.com/package/@cityofzion/neon-dappkit-types) dependencies, and also some class that implements the `Neo3-Invoker` class, like [`WalletConnect SDK`](https://github.com/CityOfZion/wallet-connect-sdk/tree/main) or [`NeonInvoker`](https://github.com/CityOfZion/neon-dappkit/blob/main/packages/neon-dappkit/src/NeonInvoker.ts) from [`@cityofzion/neon-dappkit`](https://www.npmjs.com/package/@cityofzion/neon-dappkit).

## 2. Using CPM 

> It's pretty easy to use CPM to add an off-chain SDK in your TypeScript project. After installing CPM, you can run it with a `cpm.yaml` or you could directly use the command prompt. In this tutorial let's suppose we want to easily invoke functions from the [Neo Token contract](https://dora.coz.io/contract/neo3/mainnet/0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5).

Reviewer: We don't need this "spoiler" text, instead of saying what we are gonna do, let's simply go step by step explaining them individually. And we can move the following part to the beginning explaining the objective.
> let's suppose we want to easily invoke functions from the [Neo Token contract](https://dora.coz.io/contract/neo3/mainnet/0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5).

> ### With `.yaml` file

Reviewer: We are not gonna have 2 paths .yaml OR command. Let's only focus on the more maintenable path, something I mentioned here: https://discord.com/channels/431152129789526016/1131990719272321074/1171513058577879101

Reviewer: Let's not ask the user to edit the cpm.yaml manually. Only `cpm init` and another command to add the Neo Contract (saving automatically to the cpm.yaml)

Reviewer: We need to implement this issue https://app.clickup.com/t/86a1bvure

You can generate a default `cpm.yaml` file by running `cpm init` on the terminal, it should create a file that is getting the [Props](https://github.com/CityOfZion/props) smart contract. Let's suppose the contract you want to create a SDK of is the NEO token contract, after tweaking the `cpm.yaml` file it should be something like this:

```yaml
defaults:
  contract-source-network: mainnet
  contract-generate-sdk: true
  contract-download: false
  off-chain:
    languages:
    - ts

contracts:
  - label: NEO token
    script-hash: '0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5'

networks:
  - label: mainnet
    hosts:
      - 'https://mainnet1.neo.coz.io:443'
      - 'http://seed1.neo.org:10332'
  - label: testnet
    hosts:
      - 'https://testnet1.neo.coz.io:443'
  - label: priv
    hosts:
      - 'http://127.0.0.1:10332'

```

> Then just run `cpm run` on the terminal and the SDK should be generated at the `cpm/offchain/ts/neotoken` folder. Check CPM's [documentation](https://github.com/CityOfZion/cpm/blob/master/docs/config.md) to see other options and how to better configure this `cpm.yaml` file.

Reviewer: Let's split this text in 2 parts, and since we are not asking to the reader to edit the yaml file anymore, we can simply make a small observation that he can also edit the file following the guide.

```
cpm download manifest -c 0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5 -N https://mainnet1.neo.coz.io:443
```
Then you'll need to run the generate command:
```
cpm generate ts -m .\contract.manifest.json -c 0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5
```
And the SDK will be generated at the `cpm/offchain/ts/neotoken` folder.
## 3. Initializing and running the SDK
Now that the SDK is already generated, all you need to do is import it in your project and initialize it to be able to call it.
You'll only need to instantiate a class that implements `Neo3-Invoker`, like [`NeonInvoker`](https://github.com/CityOfZion/neon-dappkit/blob/main/packages/neon-dappkit/src/NeonInvoker.ts) or [`WalletConnect SDK`](https://github.com/CityOfZion/wallet-connect-sdk/tree/main).
```ts
import { NeoToken } from './cpm_out/offchain/ts/neotoken'

import { NeonInvoker } from "@cityofzion/neon-dappkit"
import WcSdk from '@cityofzion/wallet-connect-sdk-core'
import SignClient from '@walletconnect/sign-client'

async function createNeoTokenSDK(): Promise<NeoToken> {
	return new NeoToken({
		scriptHash: NeoToken.SCRIPT_HASH,

> 		// Alternatively, use WcSDK instead of NeonInvoker

Reviewer: You can remove this comment

		})
	})
}

async function createNeoTokenSDKWithWcSDK(): Promise<NeoToken> {
	let wcSdk
	let signClient

	signClient = await SignClient.init({
        projectId: 'a9ff54e3d56a52230ed8767db4d4a810', // the ID of your project on Wallet Connect website
        relayUrl: 'wss://relay.walletconnect.com', // we are using walletconnect's official relay server
        metadata: {
          name: 'MyApplicationName', // your application name to be displayed on the wallet
          description: 'My Application description', // description to be shown on the wallet
          url: 'https://myapplicationdescription.app/', // url to be linked on the wallet
          icons: ['https://myapplicationdescription.app/myappicon.png'], // icon to be shown on the wallet
        },
    })
    wcSdk = new WcSdk(signClient)

    return new NeoToken({
		scriptHash: NeoToken.SCRIPT_HASH,
		invoker: wcSdk
	})
}

```

Then you can invoke the contract methods with this SDK:
```ts
async main() {
	const neoToken = await createNeoTokenSDKWithNeonDappkit()
	console.log(await neoToken.symbol())        // NEO
	console.log(await neoToken.decimals())      // 0
	console.log(await neoToken.totalSupply())   // 100000000
}

> ```

Reviewer: It's important to explain the difference between using NeonInvoker and WcSdk and why would you choose one over the other, the different situations where you need one over the other.

---
title: 'How to use CPM to generate an off-chain TypeScript SDK'
description: "The tutorial aims to provide a comprehensive guide for developers interested in creating TypeScript SDKs for Neo smart contracts, making it easier for them to leverage CPM and successfully work with off-chain SDKs."
author: coz.io
tags: ["TS", "CPM", "SDK"]
skill: BEGINNER
sidebar: true
---

[CPM](https://github.com/CityOfZion/cpm) is a tool that can help you download smart contracts and generate on-chain and off-chain SDKs for multiple languages without the need to write them yourself and needing none or minimal changes.

## 1. Initializing workspace
We recommend using [vite](https://vitejs.dev/guide/) to create your project from a template. You just need to run `npm create vite@latest` and select the prompts to generate your project. After creating your project, it's time to start using CPM to automate the off-chain SDK generation.

## 2. Using CPM 

> It's pretty easy to use CPM to add an off-chain SDK to your TypeScript project. After installing CPM, you can run it with a `cpm.yaml` or you could directly use the command prompt. In this tutorial, we will be using the easiest one to maintain.

Reviewer 2: ```suggestion
It's pretty easy to use CPM to add an off-chain SDK to your TypeScript project. After installing CPM you can generate a single SDK directly from the command prompt or setup a configuration file that will be processed and allows for generating multiple SDKs at the same time. In this tutorial, we will be using the easiest one to maintain.
```

The reason for my suggestion is that in the original text 
> you can run it with a `cpm.yaml`

`cpm.yaml` is mentioned for the first time in the article without having explained what it is. Not everybody might know that this is the configuration file.

> We need to generate a default `cpm.yaml` file by running `cpm init` on the terminal, it should create a file that is getting the [Props](https://github.com/CityOfZion/props) smart contract. While it is a useful contract, it's probably not the one you want, in this article, let's suppose we want to easily invoke functions from the [Neo Token contract](https://dora.coz.io/contract/neo3/mainnet/0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5).

Reviewer 2: ```suggestion
We need to generate a configuration file by running `cpm init` on the terminal. This creates a `cpm.yaml` file that by default gets the [Props](https://github.com/CityOfZion/props) smart contract. While it is a useful contract, it's probably not the one you want, in this article, let's suppose we want to easily invoke functions from the [Neo Token contract](https://dora.coz.io/contract/neo3/mainnet/0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5).
```

Slightly modified to use stronger wording on the creation. Instead of assuming it might fail to create the file, we assume it will always be successful.

> The `cpm.yaml` is divided into 4 different sections that will be tweaked to generate the desired SDK:

Reviewer 2: ```suggestion
The configuration file is divided into 4 different sections that will be tweaked to generate the desired SDK:
```

- `defaults`: will indicate that we want to generate an off-chain TypeScript SDK inside the `src` folder and that we don't want to download the NEF and manifest;

> - `contracts`: will have the Neo Token contract that we want to generate;

Reviewer 2: ```suggestion
- `contracts`: will have the Neo Token contract that we want to generate an SDK for;
```

- `tools`: since we don't want to download the smart contract, this section is unnecessary;
- `networks`: will have the Main Net network where the Neo Token contract is.

After tweaking the `cpm.yaml` file it should be something like this:

```yaml
defaults:
  contract-source-network: mainnet
  contract-generate-sdk: true
  contract-download: false
  off-chain:
    languages:
    - ts
    destinations:
      ts: src/contracts

contracts:
  - label: NEO token
    script-hash: '0xef4073a0f2b305a38ec4050e4d3d28bc40ea63f5'

networks:
  - label: mainnet
    hosts:
      - 'https://mainnet1.neo.coz.io:443'
      - 'http://seed1.neo.org:10332'
```

> For more information on how to configure `cpm.yaml`, check out [CPM's documentation](https://github.com/CityOfZion/cpm/blob/master/docs/config.md)

> Then just run `cpm run` on the terminal and the SDK should be generated at the `src/contracts/neotoken` folder.' Anytime the contract adds or removes method signatures, you'll have to run `cpm run` again to update your SDK.

Reviewer: ```suggestion
Then just run `cpm run` on the terminal and the SDK should be generated at the `src/contracts/neotoken` folder. Anytime the contract adds or removes method signatures, you'll have to run `cpm run` again to update your SDK.
```

To initialize the SDK, you'll first need to instantiate a class that implements `Neo3-Invoker`.
To enhance security and user experience, it's advisable to use [`WalletConnect`](https://github.com/CityOfZion/wallet-connect-sdk) for scenarios requiring user authentication, such as signing transactions or accessing user account details. However, for situations where such authentication is unnecessary, [`NeonInvoker`](https://github.com/CityOfZion/neon-dappkit/blob/main/packages/neon-dappkit/src/NeonInvoker.ts) from NeonDappkit offers a more streamlined solution, as it eliminates the need for users to approve requests through their wallets. In our example, we employ both methods: NeonInvoker is utilized when the user is not logged into their wallet.
```ts
import { NeoToken } from './contracts/neotoken'

import { Neo3Invoker } from "@cityofzion/neon-dappkit-types"
import { NeonInvoker } from "@cityofzion/neon-dappkit"
import WcSdk from '@cityofzion/wallet-connect-sdk-core'

> import SignClient from '@walletconnect/sign-client'
> 
> let wcSdk: WcSdk
> let neonInvoker: NeonInvoker
> 
> async function initInvokers(): Promise<NeoToken> {
>   let signClient
> 
>   signClient = await SignClient.init({
>     projectId: 'a9ff54e3d56a52230ed8767db4d4a810', // the ID of your project on Wallet Connect website
>     relayUrl: 'wss://relay.walletconnect.com', // we are using walletconnect's official relay server
>     metadata: {
>     name: 'MyApplicationName', // your application name to be displayed on the wallet
>     description: 'My Application description', // description to be shown on the wallet
>     url: 'https://myapplicationdescription.app/', // url to be linked on the wallet
>     icons: ['https://myapplicationdescription.app/myappicon.png'], // icon to be shown on the wallet
>     },
>   })
>   wcSdk = new WcSdk(signClient)

Reviewer: After fixing a [compatibility problem with signClient](https://app.clickup.com/t/86a1dpp9f) we changed the sdk initalization to pass the options directly (the older way is still possible, but depending on the signClient version, it will show type errors). [Reference](https://github.com/CityOfZion/wallet-connect-sdk/blob/main/packages/wallet-connect-sdk-core/README.md#initialize-the-client) 
```suggestion
let wcSdk: WcSdk
let neonInvoker: NeonInvoker

async function initInvokers(): Promise<NeoToken> {
  wcSdk = await WcSdk.Init({
    projectId: 'a9ff54e3d56a52230ed8767db4d4a810', // the ID of your project on Wallet Connect website
    relayUrl: 'wss://relay.walletconnect.com', // we are using walletconnect's official relay server
    metadata: {
      name: 'MyApplicationName', // your application name to be displayed on the wallet
      description: 'My Application description', // description to be shown on the wallet
      url: 'https://myapplicationdescription.app/', // url to be linked on the wallet
      icons: ['https://myapplicationdescription.app/myappicon.png'], // icon to be shown on the wallet
    },
  })
```
